### DISCUSSION TITLE
JS STORE Delete based on multiple fields

### DISCUSSION THREAD

**Discussion by Mleipper** [ujjwalguptaofficial/JsStore#320] [Q&A]
0

I have a [jsstore](https://jsstore.net/) table with the following schma

```
const loiCommonNameSchema = {
    name: 'LOICommonName',
    columns: {
        projectId: { dataType: 'string', index: true },
        Uniclass2_Id: { dataType: 'string' },
        commonName: { dataType: 'string', allowNull: true },
        code: { dataType: 'string', allowNull: true },
        cacheKey: { dataType: 'string',  index: true },
    },
};
```
I'm trying to run a delete operation on this table in the following way

```

            var dae =  await connection.remove({
                from:  'LOICommonName',
                where: {
                    projectId: projectId,
                    cacheKey: { '!=': 'hello' }
                },
                ignoreCase: true
            });
```
this is return store.worker.min.js:2 {message: "No index found for column 'undefined'. Query can not be executed without index.", type: 'no_index_found'}

I've tried to run a delete on just the projectId and this works I've also tried to run a delete on just the cacheKey and the works as expected.

I ran a select with the Where and this returned the set of data that I would expect to get deleted.

I'm not sure what I need to change in my schema to allow for a delete with a complex WHERE statement.

Any help would be much appreciated.

Many thanks ahead of time.

**Comment by ujjwalguptaofficial** [comment]
This should work clearly as its working in select query, please provide me something reproducable to fix.

But for now you can do several things  to make it work - 

* you can use array query - 

```
var dae = await connection.remove({
    from: 'LOICommonName',
    where: [
         {
            projectId: projectId
        },
        cacheKey: {
            '!=': 'hello'
        }
    ]
});
```

* you can use in query to perform this - 

do a select query and then pass ids in `in` query.

**Comment by Mleipper** [comment]
Hi, sorry to have taken a while getting back: 
I've tried the above and am still seeing {message: "No index found for column '!='. Query can not be executed without index.", type: 'no_index_found'}
```
var dae = await connection.remove({
    from: 'LOICommonName',
    where: [
         {
            projectId: projectId
        },
        cacheKey: {
            '!=': 'hello'
        }
    ]
});
```


in the code I'm declaring the table 
```
const loiCommonNameSchema = {
    name: 'LOICommonName',
    columns: {
        projectId: { dataType: 'string', index: true },
        Uniclass2_Id: { dataType: 'string' },
        commonName: { dataType: 'string', allowNull: true },
        code: { dataType: 'string', allowNull: true },
        cacheKey: { dataType: 'string',  index: true },
    },
};

};
```

```
I'm then inserting the data as such

            var testObject = {
                projectId: "testProjectId",
                Uniclass2_Id: "test Name",
                commonName: "test commonName",
               code: "test code",
               cacheKey:"test cacheKey"
            }

 var noOfRowsInserted =  await connection.insert({
                        into: "LOICommonName",
                        values: testObject , 
                    });

```


I've then got a function that I'm calling after logging the insert of the insert to delete the data. Which now reads as per your example

```
 var dae = await connection.remove({
    from: 'LOICommonName',
    where: [
         {
            projectId: projectId
        },
        cacheKey: {
            '!=': 'hello'
        }
    ]
});

```

This is where the error gets thrown whenever I attempt to delete based on multiple conditions. 
This results in a table with data in this format. 
![image](https://github.com/ujjwalguptaofficial/JsStore/assets/9982768/caac2d19-46e4-4a3a-9205-77ddba79c02b)

I've also tried a delete by selecting the objects into an array then passing the array into the delete as such 
```
                var LOICommonNameResultstest = await connection.select({
                    from: "LOICommonName",
                    projectId: projectId,
                    cacheKey: {
                        '!=': 'hello'
                    },
                    ignoreCase: true
                });

                    let loiCommonNameWhere = {
                        in: LOICommonNameResultstest
                    }

var dae = await connection.remove({
    from: 'LOICommonName',
    where: loiCommonNameWhere 
});

```

**Reply by ujjwalguptaofficial** [reply]
I can understand the issue, but i am not able to reproduce this. Could you please provide me a repo where i can reproduce and then fix in JsStore.

**Reply by ujjwalguptaofficial** [reply]
you can try reproducing in idbstudio - https://ujjwalguptaofficial.github.io/idbstudio/?db=Demo&query=select(%7B%0A%20%20%20%20%20%20from%3A'Customers'%2C%0A%20%20%20%20%20%20limit%3A100%0A%7D)

and share link with me or query.
